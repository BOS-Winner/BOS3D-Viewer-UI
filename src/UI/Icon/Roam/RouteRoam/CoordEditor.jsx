import React from 'react';
import PropTypes from "prop-types";
import style from "./style.less";

function CoordEditor(props) {
  const [x, y, z] = props.coord;
  const xRef = React.useRef(null);
  const yRef = React.useRef(null);
  const zRef = React.useRef(null);
  const onChange = React.useCallback(ev => {
    ev.preventDefault();
    ev.stopPropagation();
    const _x = Number(xRef.current.value);
    const _y = Number(yRef.current.value);
    const _z = Number(zRef.current.value);
    props.onChange([_x, _y, _z]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className={style.coordEditor} onChange={onChange}>
      <div>
        <span>X</span>
        <input type="number" value={x} ref={xRef} step={1} onChange={() => {}} />
      </div>
      <div>
        <span>Y</span>
        <input type="number" value={y} ref={yRef} step={1} onChange={() => {}} />
      </div>
      <div>
        <span>Z</span>
        <input type="number" value={z} ref={zRef} step={1} onChange={() => {}} />
      </div>
    </div>
  );
}

CoordEditor.propTypes = {
  coord: PropTypes.arrayOf(PropTypes.number).isRequired,
  onChange: PropTypes.func.isRequired,
};

export default CoordEditor;
