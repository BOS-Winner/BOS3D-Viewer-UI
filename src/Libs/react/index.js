import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
