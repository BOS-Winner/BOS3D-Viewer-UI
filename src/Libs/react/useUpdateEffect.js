import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
