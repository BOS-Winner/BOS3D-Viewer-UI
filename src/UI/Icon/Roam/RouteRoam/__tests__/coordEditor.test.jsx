import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import CoordEditor from '../CoordEditor';

const onChange = jest.fn();
let editorContainer;
let rerenderEditor;

describe('coord editor', () => {
  beforeEach(() => {
    const { container, rerender } = render(<CoordEditor coord={[1, 2, 3]} onChange={onChange} />);
    editorContainer = container;
    rerenderEditor = rerender;
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render dom', () => {
    expect(editorContainer).toMatchSnapshot();
    rerenderEditor(<CoordEditor coord={[4, 5, 6]} onChange={onChange} />);
    expect(editorContainer).toMatchSnapshot();
  });

  it('should call onChange', () => {
    onChange.mockClear();
    const inputs = editorContainer.querySelectorAll('input');
    fireEvent.change(inputs[0], {
      target: {
        value: 10,
      }
    });
    expect(onChange).lastCalledWith([10, 2, 3]);
    fireEvent.change(inputs[1], {
      target: {
        value: 10,
      }
    });
    expect(onChange).lastCalledWith([1, 10, 3]);
    fireEvent.change(inputs[2], {
      target: {
        value: 10,
      }
    });
    expect(onChange).lastCalledWith([1, 2, 10]);
    expect(onChange).toBeCalledTimes(3);
  });
});
