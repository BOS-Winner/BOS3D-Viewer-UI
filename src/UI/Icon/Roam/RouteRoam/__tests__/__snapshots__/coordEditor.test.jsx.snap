// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coord editor should render dom 1`] = `
<div>
  <div
    class="coordEditor"
  >
    <div>
      X:
      <input
        step="10"
        type="number"
        value="1"
      />
    </div>
    <div>
      Y:
      <input
        step="10"
        type="number"
        value="2"
      />
    </div>
    <div>
      Z:
      <input
        step="10"
        type="number"
        value="3"
      />
    </div>
  </div>
</div>
`;

exports[`coord editor should render dom 2`] = `
<div>
  <div
    class="coordEditor"
  >
    <div>
      X:
      <input
        step="10"
        type="number"
        value="4"
      />
    </div>
    <div>
      Y:
      <input
        step="10"
        type="number"
        value="5"
      />
    </div>
    <div>
      Z:
      <input
        step="10"
        type="number"
        value="6"
      />
    </div>
  </div>
</div>
`;
